subprojects/language-native/src/integTest/groovy/org/gradle/language/swift/SwiftCachingIntegrationTest.groovy

```diff
@@ -0,0 +1,171 @@
+            apply plugin: 'swift-executable'
+            dependencies {
+                implementation project(':hello')
+            }
+
+            project(':hello') {
+                apply plugin: 'swift-library'
+                dependencies {
+                    api project(':log')
+                }
+            }
+            project(':log') {
+                apply plugin: 'swift-library'
+            }
+        '''
+        app.logLibrary.writeToProject(project.file('log'))
+        app.library.writeToProject(project.file('hello'))
+        app.executable.writeToProject(project)
+    }
+
+    def 'compilation can be cached'() {
+        setupProject()
+
+        when:
+        withBuildCache().run compileTask(buildType)
+
+        then:
+        compileIsNotCached(buildType)
+
+        when:
+        withBuildCache().run 'clean', installTask(buildType)
+
+        then:
+        compileIsCached(buildType)
+        installation(installDir(buildType)).exec().out == app.expectedOutput
+
+        where:
+        buildType << ['Debug', 'Release']
+    }
+
+    @Unroll
+    def "compilation task is relocatable (#buildType)"() {
+        def originalLocation = file('original-location')
+        def newLocation = file('new-location')
+        setupProject(originalLocation)
+        setupProject(newLocation)
+
+        when:
+        inDirectory(originalLocation)
+        withBuildCache().run compileTask(buildType)
+
+        then:
+        compileIsNotCached(buildType)
+
+        when:
+        originalLocation.deleteDir()
+        executer.beforeExecute {
+            inDirectory(newLocation)
+        }
+
+        run 'clean'
+        withBuildCache().run compileTask(buildType), installTask(buildType)
+
+        then:
+        compileIsCached(buildType, newLocation)
+        installation(installDir(newLocation, buildType)).exec().out == app.expectedOutput
+
+        where:
+        buildType << ['Debug', 'Release']
+    }
+
+    @Unroll
+    def "downstream compilation can use cached artifacts (#buildType)"() {
+        def originalLocation = file('original-location')
+        def newLocation = file('new-location')
+        def upstreamCompileTasks = [compileTask(':hello', buildType), compileTask(':log', buildType)] as String[]
+        setupProject(originalLocation)
+        setupProject(newLocation)
+
+        when:
+        inDirectory(originalLocation)
+        withBuildCache().run compileTask(buildType), installTask(buildType)
+
+        then:
+        compileIsNotCached(buildType)
+
+        when:
+        originalLocation.deleteDir()
+        executer.beforeExecute {
+            inDirectory(newLocation)
+        }
+        withBuildCache().succeeds upstreamCompileTasks
+
+        then:
+        skipped(upstreamCompileTasks)
+
+        when:
+        run compileTask(buildType), installTask(buildType)
+
+        then:
+        skipped(upstreamCompileTasks)
+        compileIsNotCached(buildType)
```

Comment:
I think mixing the general purpose `skipped()` and the specific `compileIsNotCached()` should instead be symmetric. How about `executedAndNotSkipped(compileTask(buildType))`? Or use a local variable similar to `upstreamCompileTask` for `compileTask` and check that?

Reply:
I removed the `compileIsNotCached` and `compileIsCached` methods.